mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp

```diff
@@ -0,0 +1,94 @@
+//===- MemRefToEmitC.cpp - MemRef to EmitC conversion ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert memref ops into emitc ops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/MemRefToEmitC/MemRefToEmitC.h"
+
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+namespace {
+struct ConvertAlloca final : public OpConversionPattern<memref::AllocaOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(memref::AllocaOp op, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (!op.getType().hasStaticShape()) {
+      return rewriter.notifyMatchFailure(
+          op.getLoc(), "cannot transform alloca with dynamic shape");
+    }
+
+    if (op.getAlignment().value_or(1) > 1) {
+      // TODO: Allow alignment if it is not more than the natural alignment
+      // of the C array.
+      return rewriter.notifyMatchFailure(
+          op.getLoc(), "cannot transform alloca with alignment requirement");
+    }
+
+    auto resultTy = getTypeConverter()->convertType(op.getType());
+    auto noInit = emitc::OpaqueAttr::get(getContext(), "");
+    rewriter.replaceOpWithNewOp<emitc::VariableOp>(op, resultTy, noInit);
+    return success();
+  }
+};
+
+struct ConvertLoad final : public OpConversionPattern<memref::LoadOp> {
```

Comment:
Hm, this is the place where the lvalue-to-rvalue op on the result of the emitc.subscript is needed to perform the actual read.
How close are you to opening a PR for that?

Reply:
From my understanding, the lvale-to-rvalue op is nothing that can be upstreamed in its current shape. It's rather a prototype which builds a basis to discuss how to handle this issue. However, it might not be the solution we want to have upstream. From my last offline chat with @simon-camp, there will soon be more details on discourse to stimulate further discussions but I wouldn't expect a PR soon.

Reply:
I could add an `emitc.read` op into this PR which will (only) take an `emitc.subscript` and materialize it into a local variable (i.e. `var v7 = v1[v2][v3];`). This would address the immediate problem and could later be morphed into a more general lvalue-to-rvalue op as soon as we pinned that down.
What do you think?

Reply:
Generally speaking we try to avoid introducing ops while already intending to replacing them near term. Right now, I am not enough into this to make a statement on what other options we have and if an `emitc.read` would solve the problem. @simon-camp can you comment on this?

Reply:
Re-reading @simon-camp's initially message, I see now how `variable+subscript+assign` would be a good workaround until we have something for lvalues. I will try to go with that approach.